lib/custom/src/MShop/Service/Provider/Payment/PaypalPlus.php

```diff
@@ -0,0 +1,227 @@
+            'default' => '0',
+            'required' => false,
+        ),
+        'testmode' => array(
+            'code' => 'testmode',
+            'internalcode' => 'testmode',
+            'label' => 'Test mode without payments',
+            'type' => 'boolean',
+            'internaltype' => 'boolean',
+            'default' => '1',
+            'required' => true,
+        ),
+    );
+    private $provider;
+
+    public function getConfigBE(): array
+    {
+        $list = [];
+
+        foreach ($this->beConfig as $key => $config) {
+            $list[$key] = new \Aimeos\MW\Criteria\Attribute\Standard($config);
+        }
+
+        return $list;
+    }
+
+    protected function getProvider(): \Omnipay\Common\GatewayInterface
+    {
+        if (!isset($this->provider)) {
+            $this->provider = OPay::create('PayPal_Rest');
+            $this->provider->setTestMode((bool) $this->getValue('testmode', false));
+            $this->provider->initialize($this->getServiceItem()->getConfig());
+        }
+
+        return $this->provider;
+    }
+
+    protected function getPaymentForm(\Aimeos\MShop\Order\Item\Iface $order, array $params): \Aimeos\MShop\Common\Helper\Form\Iface
+    {
+        $list = [];
+        $baseItem = $this->getOrderBase($order->getBaseId(), \Aimeos\MShop\Order\Item\Base\Base::PARTS_ADDRESS);
+        $addresses = $baseItem->getAddress(\Aimeos\MShop\Order\Item\Base\Address\Base::TYPE_PAYMENT);
+
+        $parts = \Aimeos\MShop\Order\Item\Base\Base::PARTS_SERVICE
+        | \Aimeos\MShop\Order\Item\Base\Base::PARTS_PRODUCT
+        | \Aimeos\MShop\Order\Item\Base\Base::PARTS_ADDRESS;
+
+        $base = $this->getOrderBase($order->getBaseId(), $parts);
+        $data = $this->getData($base, $order->getId(), $params);
+        $urls = $this->getPaymentUrls();
+        try {
+            $response = $this->sendRequest($order, $data);
+            $approvalUrl = "" ;
+            $testmode = "" ;
+            $countryid = "" ;
+            if ($response->isSuccessful()) {
+                $this->setOrderData($order, ['Transaction' => $response->getTransactionReference()]);
+                $this->saveRepayData($response, $base->getCustomerId());
+                $this->saveOrder($order->setPaymentStatus(Status::PAY_UNFINISHED));
+
+                if (!empty($response->getData()['links']['1']['href'])) {
+                    $approvalUrl = $response->getData()['links']['1']['href'] ;
+                }
+
+                $addresses = $base->getAddress(\Aimeos\MShop\Order\Item\Base\Address\Base::TYPE_PAYMENT);
+
+                if (($address = current($addresses)) !== false) {
+                    $countryid = $address->getCountryId();
+                }
+
+                if (empty($approvalUrl) || empty($countryid)) {
+                    throw new \Aimeos\MShop\Service\Exception($response->getMessage());
+                }
+            } elseif ($response->isRedirect()) {
```

Comment:
A redirect isn't possible, isn't it?

Reply:
yes it is , but I kept it as I had copy updateSync() from onmipay.php ..
so ,no need for it and I will remove it